src/axom/sidre/core/ItemCollection.hpp

```diff
@@ -150,43 +151,130 @@ class ItemCollection
   ///
   virtual IndexType getNextValidIndex(IndexType idx) const = 0;
 
-  ///
-  virtual bool hasItem(const std::string& name) const = 0;
-
   ///
   virtual bool hasItem(IndexType idx) const = 0;
 
   ///
-  virtual TYPE* getItem(const std::string& name) = 0;
+  virtual T* getItem(IndexType idx) = 0;
 
   ///
-  virtual TYPE const* getItem(const std::string& name) const = 0;
+  virtual T const* getItem(IndexType idx) const = 0;
 
   ///
-  virtual TYPE* getItem(IndexType idx) = 0;
+  virtual IndexType insertItem(T* item, const std::string& name = "") = 0;
 
   ///
-  virtual TYPE const* getItem(IndexType idx) const = 0;
+  virtual T* removeItem(IndexType idx) = 0;
 
   ///
-  virtual const std::string& getItemName(IndexType idx) const = 0;
+  virtual void removeAllItems() = 0;
 
-  ///
-  virtual IndexType getItemIndex(const std::string& name) const = 0;
+public:
+  class iterator : public IteratorBase<iterator, IndexType>
+  {
+  private:
+    using BaseType = IteratorBase<iterator, IndexType>;
+    using CollectionType = ItemCollection<T>;
 
-  ///
-  virtual IndexType insertItem(TYPE* item, const std::string& name) = 0;
+  public:
+    // Iterator traits required to satisfy LegacyRandomAccessIterator concept
+    // before C++20
+    // See: https://en.cppreference.com/w/cpp/iterator/iterator_traits
+    using difference_type = IndexType;
+    using value_type = typename std::remove_cv<T>::type;
+    using reference = T&;
+    using pointer = T*;
+    using iterator_category = std::forward_iterator_tag;
 
-  ///
-  virtual TYPE* removeItem(const std::string& name) = 0;
+  public:
+    iterator(CollectionType* coll, bool is_first) : m_collection(coll)
+    {
+      SLIC_ASSERT(coll != nullptr);
 
-  ///
-  virtual TYPE* removeItem(IndexType idx) = 0;
+      BaseType::m_pos =
+        is_first ? coll->getFirstValidIndex() : sidre::InvalidIndex;
+    }
 
-  ///
-  virtual void removeAllItems() = 0;
+    pointer operator->() { return m_collection->getItem(BaseType::m_pos); }
+
+    reference operator*() { return *m_collection->getItem(BaseType::m_pos); }
+
+  private:
+    // Remove backwards iteration functions
+    using BaseType::operator--;
+    using BaseType::operator-=;
```

Comment:
Is there a reason these have to be forward iterators, and not random-access iterators?

Reply:
Just asking out of curiosity, although I think there are some STL algorithms that require random-access iterators.

But doesn't sound like it'd be worth it if we can't implement it in a straightforward manner (also, we'd need to do it in O(1) to comply with the standard).